### ISSUE TITLE
How to best add support for IDENTIFIER() clause

### ISSUE THREAD

**Issue by kacpermuda** [apache/datafusion-sqlparser-rs#1412]
I am thinking about adding support for the IDENTIFIER() clause (available in Snowflake and Databricks, see docs below). 

In OpenLineage, I was able to [partially support it](https://github.com/OpenLineage/OpenLineage/pull/2999) by implementing visitors, but that only worked for a couple of statement types (SELECT, MERGE, UPDATE, DELETE) - the ones where i was able to access a TableFactor and not only ObjectName. 

I believe, that in order to be able to recognize every use case of IDENTIFIER clause in the visitors I'd have to modify all the statements to provide access to TableFactor or some similar struct. This specific clause if problematic as it can appear anywhere an object name is referenced, and i believe there is no single place in the parser where i could add support for such a syntax. For now I am only interested in recognizing this clause in table names, so that limits the scope a little bit.

Does anyone have some ideas on how to best approach this problem? I'd like to use the limited resources I have in the best possible way and I feel like I'm missing something obvious here.

@iffyio, maybe you have some time and idea for that? You helped me with my previous PR 😄 

Docs:
[Snowflake](https://docs.snowflake.com/en/sql-reference/identifier-literal)
[Databricks](https://docs.databricks.com/en/sql/language-manual/sql-ref-names-identifier-clause.html)

**Comment by jmhain** [comment]
I think the cleanest way would be to introduce an `enum` to replace `ObjectName` wherever an `IDENTIFIER` clause is allowed, something like:

```rust
enum ObjectNameOrIdentifierClause  {
    Name(ObjectName),
    IdentifierClause(IdentifierClause)
}

enum IdentifierClause {
    String(...),
    SessionVar(...),
    ...
}
```

**Comment by iffyio** [comment]
Yeah I think @jmhain's suggestion to use the enum sounds good to support this!

**Comment by iffyio** [comment]
Related PR https://github.com/apache/datafusion-sqlparser-rs/pull/1539

**Comment by LucaCappelletti94** [comment]
Completed by #1539.
